Wikipedia/Code/WMFAppViewController+Extensions.swift

```diff
@@ -97,13 +101,119 @@ extension WMFAppViewController: NotificationsCenterPresentationDelegate {
+
+        while let presentedViewController = currentController?.presentedViewController {
+            if let presentedNavigationController = (presentedViewController as? UINavigationController) {
+                for viewController in presentedNavigationController.viewControllers {
+                    if let editingFlowViewController = viewController as? EditingFlowViewController {
+                        return editingFlowViewController
+                    }
+                }
+            } else if let editingFlowViewController = presentedViewController as? EditingFlowViewController {
+                return editingFlowViewController
+            }
+            
+            currentController = presentedViewController
+        }
+
+        return nil
+    }
+    
+    private var topMostViewController: UIViewController? {
+            
+        var topViewController: UIViewController = navigationController ?? self
+
+        while let presentedViewController = topViewController.presentedViewController {
+            topViewController = presentedViewController
+        }
+
+        return topViewController
     }
+    
+    private func presentEditorAlert(on viewController: UIViewController, confirmationBlock: @escaping () -> Void) {
+        
+        let title = CommonStrings.editorExitConfirmationTitle
+        let message = CommonStrings.editorExitConfirmationBody
+        
+        let alertController = UIAlertController(title: title, message: message, preferredStyle: .alert)
+        let discardAction = UIAlertAction(title: CommonStrings.discardEditsActionTitle, style: .destructive) { _ in
+            confirmationBlock()
+        }
+        let cancelAction = UIAlertAction(title: CommonStrings.cancelActionTitle, style: .cancel)
+        
+        alertController.addAction(discardAction)
+        alertController.addAction(cancelAction)
+        
+        viewController.present(alertController, animated: true, completion: nil)
+        
+    }
+}
+
+fileprivate extension UIViewController {
+    
+    /// Returns self or embedded view controller (if self is a UINavigationController) if conforming to NotificationsCenterFlowViewController
+    /// Does not consider presenting view controllers
+    var notificationsCenterFlowViewController: NotificationsCenterFlowViewController? {
+        
+        if let viewController = self as? NotificationsCenterFlowViewController {
+            return viewController
+        }
+        
+        if let navigationController = self as? UINavigationController,
+           let viewController = navigationController.viewControllers.last as? NotificationsCenterFlowViewController {
+            return viewController
+        }
+
+        return nil
+    }
+}
+
+
+/// View Controllers that have an editing element (Section editor flow, User talk pages, Article description editor)
+protocol EditingFlowViewController where Self: UIViewController {
```

Comment:
Is the `where Self: UIViewController` conditional conformance actually needed here?

Reply:
@staykids I don't think it's needed to compile, but I liked constraining them to `UIViewController` because I am only considering UIKit view controllers when looking around in the navigation hierarchy for these types (within `editingFlowViewControllerInHierarchy` and `notificationsCenterFlowViewController`. I did initially try an option with no constraint and conformed the `NotificationsCenterFilterView` and `NotificationsCenterInboxView` SwiftUI structs to it, but I had a hard time checking for that conformance from WMFAppViewController. I may just be missing something with SwiftUI / UIKit interop though.